### ISSUE TITLE
How to use MultipleNegativesRankingLoss?

### ISSUE THREAD

**Issue by pommedeterresautee** [UKPLab/sentence-transformers#100]
There is no example on how to use this loss.
According to https://github.com/UKPLab/sentence-transformers/issues/9, it seems to be an interesting one for similarity search.

My question is what reader to use. From my understanding, the `label` field is not used. So anything work. Trainset should only be made of positive examples. 
Am I right?

And the way to call it looks like:

```python
...
    train_loss = losses.MultipleNegativesRankingLoss(sentence_embedder=model)
...
    model.fit(train_objectives=[(train_data_loader, train_loss)],
              evaluator=evaluator,
              epochs=opt.epoch,
              evaluation_steps=opt.evaluation_step,
              warmup_steps=warmup_steps,
              output_path=opt.output_model+"_"+datetime.now().strftime("%Y-%m-%d_%H-%M-%S"))
```
Is it right? Would it possible to push an example of use case somewhere? (when is it a useful loss, etc.).

**Comment by nreimers** [comment]
Hi @pommedeterresautee 
Correct, the label is not used. Your code looks right. 

I think in the version pypi is still a bug in MultipleNegativesRankingLoss which is fixed in the github code here.

The data has to look like this:
anchor1 positive1
anchor2 positive2
anchor3 positive3
....

AnchorX and PositiveX are considered a positive example, while all other AnchorX and PositiveY (X != Y) are considered negative examples. So you must ensure that there are no other positive examples in your batch.

I for example use MultipleNegativesRankingLoss with aligned sentences from machine translation to create multi-lingual models. One sentence pair consists of the original (English) sentence and the translated sentence. Only this pair is considered as correct translation, all other pairs are incorrect pairs.

Best
Nils Reimers

**Comment by pommedeterresautee** [comment]
Thank you for your answer. 
I have forbidden to have 2 times the same Anchor, so the dataset is Ok.
I used the STS reader, and all scores are `1`.

I run that code and found 2 issues regarding the evaluator:
- EmbeddingSimilarityEvaluator : crash because Pearson computation crash and tells me that one of the array is constant (it is right, and I agree, it can't work as true label is always 1, so no way to compute a correlation)
- BinaryEmbeddingSimilarityEvaluator: all scores are the same `1,2500,0.5,0.5,0.5` in `binary_similarity_evaluation_results.csv`. I never used that evaluator and I may be wrong.

Can you tell me which reader and which evaluator I should use?

Kind regards,
Michael

**Comment by nreimers** [comment]
Hi Michael,
using the STS reader and setting all labels to 1 works.

Currently there is no suitable evaluator to evaluate the MultipleNegativeRankingLoss. EmbeddingSimilarityEvaluator and BinaryEmbeddingSimilarityEvaluator do not work, as they require labels (and all labels are 1 in your dataset).


You evaluate  MultipleNegativeRankingLoss in the following way (using Precision@k with n negative examples):

You have n sentence pairs (a, b), e.g., n=1000. You know that a_i, b_i are correct while a_i, a_j (i!=j) is incorrect.

For all 2x1000 sentences you generate the sentence embedding and then check if cos(a_i, b_i) >= cos(a_i, b_j) for all i and j. This would be Precision@1.

To formulate this less mathematically: For each a_i you check whether from all b's, b_i has the highest similarity to a_i. If yes, then this is considered correct. If some b_j has a higher similarity to a_i than b_i, then you consider this as an error.

This gives you a score between 0% and 100%. When reaching 100%, all pairs are correctly mapped, i.e., a_i is always most similar to b_i.


I'm currently working on implementing this evaluator and hope I can share it soon.

Best
Nils Reimers

**Comment by pommedeterresautee** [comment]
Thanks a lot, it s very clear. 
I close for now the issue.

**Comment by AnonymousResearch23** [comment]
@nreimers 
Any update about MultipleNegativeRankingEvaluator?
